### ISSUE TITLE
COLMAP point clouds dosen't match

### ISSUE THREAD

**Issue by derrick-xwp** [colmap/colmap#1025]
I have used colmap to get camera parameters and imga data.

Then i want to reconstruct with external depth maps.

But the point cloud from diffferent depth map didn't match with each other.

There seems issue with transition matrix.

Why does this happen?

**Comment by tsattler** [comment]
Can you give some more details on your setup? How do you obtain the depth maps?

Structure-from-Motion systems such as Colmap are not able to recover the size of a scene, i.e., the distance between two images in a reconstruction will not be defined in meters or any other meaningful unit. If your depth maps do not align with the poses that you get from Colmap, I would assume that this is the underlying issue (but it is hard to tell without any information about what you mean by "depth map didn't match with each other"). If this is the case, you will need to scale your Colmap reconstruction such that it is consistent with the scales of the depth maps (or the other way around).

**Comment by derrick-xwp** [comment]
i run colmap with 3 color photos in the same scene.
Then i can get camera pose.
I have three depth maps corresponding to the 3 color images.
I project them to the world coordinate by the extrinsics and intrinsics from colmap results.
But from the point cloud, the three point clouds dosen't match. Seems issue with transition.
The image of point clouds is attached.
![图片](https://user-images.githubusercontent.com/53371723/96993654-2e21ac80-155e-11eb-8825-ae35f5dde9b3.png)
![图片](https://user-images.githubusercontent.com/53371723/96993672-34178d80-155e-11eb-8f38-84c30bf35cbb.png)

**Comment by derrick-xwp** [comment]
From my understanding, these three images are reconstructed in the same colmap project. So they should merge into one point cloud. the rotation seems correct, but problems with transition,
Thanks!

**Comment by tsattler** [comment]
I am still not sure whether I understand what you are doing. You obtain the depth maps with Colmap using the same sparse reconstruction?
If yes, are you sure that you are applying the correct transformation? Colmap stores poses that map from the global coordinate system into the image coordinate system.

**Comment by derrick-xwp** [comment]
> 
> 
> I am still not sure whether I understand what you are doing. You obtain the depth maps with Colmap using the same sparse reconstruction?
> If yes, are you sure that you are applying the correct transformation? Colmap stores poses that map from the global coordinate system into the image coordinate system.

no. i used depth map generated from somewhere else.
But from the result we can see that the depth map work well.
those merged point clouds are transformed by the c2w extrinsics from image to global.
So i think they should work..
Could you give me example code which applied the transformation from image to world coordinate with data from colmap?
Appreciate!

**Comment by derrick-xwp** [comment]
The geometric depth map generated by colmap is rgb image.
I have converted it to grap image by cv2 then project it to global coordinate, but the result is not good.
Do you know how can i use the geometric depth map generated in dense reconstruction ?
Thanks!

**Comment by tsattler** [comment]
> no. i used depth map generated from somewhere else.

Are you sure that the scale of the depth maps matches the scale of the Colmap reconstruction?

> The geometric depth map generated by colmap is rgb image.
I have converted it to grap image by cv2 then project it to global coordinate, but the result is not good.

Did you look at the file format of the depth maps? You can find more details here: https://colmap.github.io/format.html#depth-and-normal-maps

**Comment by ookey** [comment]
Hello,
I've got a similar problem, thus I put my question here. Please tell me if I should start another thread.
I used the cameras.txt and images.txt resp to get the intrinsics and extrinsics.
I imported geometric depthmaps from colmap. I use intrinsics, then extrinsics to get the corresponding point cloud in absolute referential.
They match "more or less" but with a high level of error (2-3cm offset as far as I can tell looking in meshlab). 
It surprises me since the reconstructed dense point cloud (stereo step of colmap) is clean.
Below the code I use.

Do you have any clue to give me?

```
import read_write_dense # in colmap/scripts/python
import numpy as np
from PIL import Image
import argparse
import os, sys
from pyquaternion import quaternion

FAR = 7.
NEAR = 1.

def parse_images(path):
    img_lines = [l[:-1] for l in open(path) if not l.startswith('#')]
    # line ex: 25 0.914934 0.0645191 0.397207 0.0309901 -3.61667 -0.843549 5.25552 25 DSC-RX0M2_5788485_DSC00009.JPG
    img_lines = img_lines[::2] # keep one line out of two (see file format in colmap doc)
    img_lines = np.array([ l.split(' ') for l in img_lines]) # array with [img_id, QW, QX, QY, QZ, TX, TY, TZ, cam_id, img_name]
    img_ids = img_lines[:,0].astype(np.int) - 1
    img_lines[img_ids] = img_lines # reorder by id
    images = img_lines[:,-1]
    camids = img_lines[:,-2].astype(np.int) - 1
    w2c_poses = np.array([quaternion.Quaternion(q).transformation_matrix for q in np.array(img_lines[:,1:5]).astype(np.float32)])
    translations = np.array(img_lines[:,5:8]).astype(np.float32)
    w2c_poses[:,:3,3] = translations
    return images, camids, w2c_poses

def parse_cameras(path):
    cam_lines = np.array([l[:-1].split(' ') for l in open(path) if not l.startswith('#')])
    # line ex: 1 PINHOLE 4800 3200 3327.18 3347.83 2400 1600
    ids = cam_lines[:,0].astype(np.int) - 1
    w_h_fx_fy_cx_cy = cam_lines[:,2:8].astype(np.float32)
    intrinsics = np.array([[[fx/h,0,cx/w],[0,fy/h,cy/h],[0,0,1]] for w,h,fx,fy,cx,cy in w_h_fx_fy_cx_cy])
    intrinsics[ids] = intrinsics # reorder by id
    return intrinsics

if __name__=='__main__':

    parser = argparse.ArgumentParser()
    parser.add_argument('--colmap_export_dir', type=str)
    parser.add_argument('--colmap_dense_dir', type=str)
    args = parser.parse_args()

    images, camids, w2c_poses = parse_images('%s/images.txt' % args.colmap_export_dir)
    intrinsics = parse_cameras('%s/cameras.txt' % args.colmap_export_dir)

    for i,(cam_id,img_name) in enumerate(zip(camids, images)):
        d = read_write_dense.read_array(os.path.join(args.colmap_dense_dir, 'stereo/depth_maps/%s.geometric.bin' % img_name))[:,:,np.newaxis]
        h,w,_ = d.shape
        print('h,w:', h, w)
        intr = intrinsics[cam_id]
        uv1 = np.array([[[(u+0.5)/w,(v+0.5)/h,1.] for u in range(w)] for v in range(h)])
        xyz_over_z = np.matmul(np.linalg.inv(intr)[np.newaxis,np.newaxis,:,:], uv1[:,:,:,np.newaxis])[:,:,:,0]
        xyz = xyz_over_z * d

        xyz = xyz.reshape(-1,3) # shape (h*w,3)
        xyz = np.matmul(np.linalg.inv(w2c_poses[i])[np.newaxis,:,:],np.concatenate([xyz,np.ones((xyz.shape[0],1))], axis=1).reshape(xyz.shape[0],4,1))[:,:3,0]
        c = np.array(Image.open('%s/images/%s' % (args.colmap_dense_dir, img_name)).resize((w,h))).reshape(-1,3) # shape (h*w,3)
        np.savetxt('test%03d.obj' % i, np.concatenate([xyz,c], axis=1), fmt='v %f %f %f %d %d %d')
```

**Comment by derrick-xwp** [comment]
hey&nbsp;&nbsp;


you mean you are using the output of colmap to reconstruct point cloud by yourself but there are misalignment?&nbsp;


But the fused. ply&nbsp; looks clean?&nbsp;




I think I have the same confusion



---Original---

**Comment by tsattler** [comment]
@ookey How are you setting the `colmap_export_dir` path? Is this the `sparse/` folder in the dense reconstruction folder? Note that as part of the dense reconstruction process, the images are undistorted, which changes their intrinsic calibration to a pinhole model and also changes their image size. If you are using the model for the original images, then the intrinsics will not be consistent and you will end up with misalignments.

**Comment by ookey** [comment]
Thanks for your answer @tsattler,
Actually, I've just re-tested with small square images (1024x1024), and my misalignment problem "disappeared" :-)
I guess my problem was something like:
* I used intrinsics got with "File > export model as text", corresponding with 2400x3200 images
* I used 2000x1333 depth_maps produced by the stereo step (and I resized the color accordingly)
I should have scaled focals and center I guess? Or is there a way to directly get the intrinsics matching the depth_maps directly?

I take the opportunity to congratulate the whole Colmap team. It's a great tool!
